File: services/api/src/mangai_api/i18n.py
from __future__ import annotations

import re

from mangai_api.constants import SUPPORTED_TEXT_DIRECTIONS, SUPPORTED_UI_LOCALES


LANGUAGE_TAG_RE = re.compile(r"^[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*$")
RTL_PRIMARY_LANGUAGES = {"ar", "fa", "he", "ur"}


class LocaleValidationError(ValueError):
    """Raised when locale or language normalization fails."""


def normalize_language_tag(value: str) -> str:
    raw = value.strip()
    if not raw:
        raise LocaleValidationError("Language tag cannot be empty.")
    if not LANGUAGE_TAG_RE.fullmatch(raw):
        raise LocaleValidationError("Invalid BCP 47-like language tag.")

    segments = raw.split("-")
    primary = segments[0].lower()
    normalized_segments = [primary]

    for segment in segments[1:]:
        if len(segment) == 2 and segment.isalpha():
            normalized_segments.append(segment.upper())
        else:
            normalized_segments.append(segment)

    return "-".join(normalized_segments)


def normalize_ui_locale(value: str) -> str:
    normalized = normalize_language_tag(value)
    if normalized not in SUPPORTED_UI_LOCALES:
        raise LocaleValidationError(
            f"Unsupported UI locale. Expected one of: {', '.join(SUPPORTED_UI_LOCALES)}."
        )
    return normalized


def infer_text_direction(language_tag: str) -> str:
    normalized = normalize_language_tag(language_tag)
    primary_language = normalized.split("-")[0]
    direction = "rtl" if primary_language in RTL_PRIMARY_LANGUAGES else "ltr"
    if direction not in SUPPORTED_TEXT_DIRECTIONS:
        raise LocaleValidationError("Unsupported inferred text direction.")
    return direction

File: packages/shared/src/i18n.ts
import { RTL_LANGUAGE_CODES, SUPPORTED_UI_LOCALES } from "./enums.ts";
import { ValidationError } from "./errors.ts";

export function canonicalizeLanguageTag(value: unknown, path: Array<string | number> = []): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new ValidationError("Expected a non-empty language tag", path);
  }

  try {
    return Intl.getCanonicalLocales(value.trim())[0]!;
  } catch {
    throw new ValidationError("Invalid BCP 47 language tag", path);
  }
}

export function normalizeUiLocale(
  value: unknown,
  path: Array<string | number> = [],
): (typeof SUPPORTED_UI_LOCALES)[number] {
  const locale = canonicalizeLanguageTag(value, path);
  if (!SUPPORTED_UI_LOCALES.includes(locale as (typeof SUPPORTED_UI_LOCALES)[number])) {
    throw new ValidationError(
      `Unsupported UI locale. Expected one of: ${SUPPORTED_UI_LOCALES.join(", ")}`,
      path,
    );
  }
  return locale as (typeof SUPPORTED_UI_LOCALES)[number];
}

export function inferTextDirectionForLanguage(languageTag: unknown): "ltr" | "rtl" {
  const normalized = canonicalizeLanguageTag(languageTag);
  const primaryLanguage = normalized.split("-")[0]!.toLowerCase();
  return RTL_LANGUAGE_CODES.includes(primaryLanguage as (typeof RTL_LANGUAGE_CODES)[number])
    ? "rtl"
    : "ltr";
}
